src/guards/index.ts

```diff
@@ -0,0 +1,86 @@
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+
+---------------------------------------------------------------------------*/
+
+import {
+  Kind,
+  TAny,
+  TAnySchema,
+  TArray,
+  TBoolean,
+  TConstructor,
+  TFunction,
+  TInteger,
+  TLiteral,
+  TNull,
+  TNumber,
+  TObject,
+  TPromise,
+  TRecord,
+  TRef,
+  TSchema,
+  TSelf,
+  TString,
+  TTuple,
+  TUint8Array,
+  TUndefined,
+  TUnion,
+  TUnknown,
+  TVoid,
+} from '../typebox'
+
+export function isTSchema(value: object): value is TSchema {
+  return Kind in value
+}
+
+type TypeGuard<T extends TSchema> = (value: object) => value is T
+
+function typeGuardFactory<T extends TAnySchema>(kind: T[typeof Kind]): TypeGuard<T> {
+  return (value): value is T => (value as any).Kind === kind
+}
+
+export const isTAny = typeGuardFactory<TAny>('Any')
+export const isTArray = typeGuardFactory<TArray>('Array')
+export const isTBoolean = typeGuardFactory<TBoolean>('Boolean')
+export const isTConstructor = typeGuardFactory<TConstructor>('Constructor')
+export const isTFunction = typeGuardFactory<TFunction>('Function')
+export const isTInteger = typeGuardFactory<TInteger>('Integer')
+export const isTLiteral = typeGuardFactory<TLiteral>('Literal')
+export const isTNull = typeGuardFactory<TNull>('Null')
+export const isTNumber = typeGuardFactory<TNumber>('Number')
+export const isTObject = typeGuardFactory<TObject>('Object')
+export const isTPromise = typeGuardFactory<TPromise>('Promise')
+export const isTRecord = typeGuardFactory<TRecord>('Record')
+export const isTSelf = typeGuardFactory<TSelf>('Self')
+export const isTRef = typeGuardFactory<TRef>('Ref')
+export const isTString = typeGuardFactory<TString>('String')
+export const isTTuple = typeGuardFactory<TTuple>('Tuple')
+export const isTUndefined = typeGuardFactory<TUndefined>('Undefined')
+export const isTUnion = typeGuardFactory<TUnion>('Union')
+export const isTUint8Array = typeGuardFactory<TUint8Array>('Uint8Array')
+export const isTUnknown = typeGuardFactory<TUnknown>('Unknown')
+export const isTVoid = typeGuardFactory<TVoid>('Void')
```

Comment:
I think this is maybe ok for now. But would be ideal if each guard checked the expected schema structure for each type. So for example.

```typescript
// This is the definition for a String
export interface TString extends TSchema, StringOptions<string> {
  [Kind]: 'String'
  static: string
  type: 'string'
}

// So the guard should check all the minimum required properties for it to be a `string`.
function isTString(schema: any) {
    return schema[Kind] === 'String' && schema.type === 'string'
}
```
Having this would essentially formalize the schemas generated by TypeBox.

Reply:
IMO this is doesn't provide any real value for users while introducing more complexity & increasing the performance cost of these functions. Checking for the `Kind` symbol should be sufficient, users creating their own schema objects with a `Kind` symbol should be expected to properly conform to the interface defined in the types.